### PR TITLE
add user entity and authentication controller

### PR DESCRIPTION
**Pull request by nikitaabo**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -93,6 +93,12 @@
 			<artifactId>springdoc-openapi-starter-webmvc-ui</artifactId>
 			<version>2.1.0</version>
 		</dependency>
+		<dependency>
+			<groupId>commons-beanutils</groupId>
```
Comment (Elena-Bruyako) [inline pom.xml]: remove

File: src/main/java/com/example/spring/controller/AuthenticationController.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/auth")
+public class AuthenticationController {
+    private final UserService userService;
+
+    @PostMapping("/registration")
+
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/spring/controller/AuthenticationController.java]:
```suggestion
```

File: src/main/java/com/example/spring/exception/CustomGlobalExceptionHandler.java [hunk 3]
```diff
@@ -41,4 +42,10 @@ private String getErrorMessage(ObjectError e) {
         }
         return e.getDefaultMessage();
     }
+
+    @ExceptionHandler(RegistrationException.class)
+    public ResponseEntity<Object> handleRegistrationException(RegistrationException ex) {
+        ErrorResponse errorResponse = new ErrorResponse("Registration failed", ex.getMessage());
+        return new ResponseEntity<>(errorResponse, HttpStatus.BAD_REQUEST);
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/spring/exception/CustomGlobalExceptionHandler.java]:
```suggestion
        return new ResponseEntity<>(errorResponse, HttpStatus.CONFLICT);
```

File: src/main/java/com/example/spring/exception/ErrorResponse.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+public class ErrorResponse {
+    private String error;
+    private String message;
+
+    public ErrorResponse(String error, String message) {
+        this.error = error;
+        this.message = message;
+    }
+
+    // Геттеры и сеттеры
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/spring/exception/ErrorResponse.java]: you can use Lombok annotations, for constructor also

File: src/main/java/com/example/spring/models/User.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+@Table(name = "users")
+@Getter
+@Setter
+@SQLDelete(sql = "UPDATE users SET is_deleted = true WHERE id=?")
+@SQLRestriction(value = "is_deleted=false")
+public class User {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Column(nullable = false)
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/spring/models/User.java]:
```suggestion
    @Column(nullable = false, unique = true)
```

File: src/main/java/com/example/spring/services/UserServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+@Service
+@RequiredArgsConstructor
+public class UserServiceImpl implements UserService {
+    private final UserRepository userRepository;
+    private final UserMapper userMapper;
+
+    @Override
+    public UserResponseDto register(UserRegistrationRequestDto registrationRequestDto)
+            throws RegistrationException {
+        Optional<User> userFromDb = userRepository.findByEmail(registrationRequestDto.getEmail());
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/spring/services/UserServiceImpl.java]:
```suggestion
        if (userRepository.existsByEmail(registrationRequestDto.email())) {
```

File: src/main/java/com/example/spring/services/UserServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+        if (userFromDb.isPresent()) {
+            throw new RegistrationException("User with email " + registrationRequestDto.getEmail()
+                    + " already exists");
+        }
+        return userMapper.toDto(userRepository.save(userMapper.toModel(registrationRequestDto)));
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/spring/services/UserServiceImpl.java]: bad practice to call methods inline
```suggestion
        User user = userMapper.toModel(registrationRequestDto);
        return userMapper.toDto(userRepository.save(user));
```

File: src/main/java/com/example/spring/validation/FieldMatchValidator.java [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    @Override
+    public boolean isValid(Object value, ConstraintValidatorContext constraintValidatorContext) {
+        try {
+            String firstFieldValue = BeanUtils.getProperty(value, firstField);
+            String secondFieldValue = BeanUtils.getProperty(value, secondField);
+
+            return firstFieldValue != null && firstFieldValue.equals(secondFieldValue);
+        } catch (Exception e) {
+            return false;
+        }
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/spring/validation/FieldMatchValidator.java]:
```suggestion
        Object field = new BeanWrapperImpl(value).getPropertyValue(this.firstField);
        Object fieldMatch = new BeanWrapperImpl(value).getPropertyValue(this.secondField);
        return Objects.equals(field, fieldMatch);
```

File: src/main/resources/db/changelog/changes/02-create-users-table.yaml [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+                  type: BIGINT
+                  autoIncrement: true
+                  constraints:
+                    primaryKey: true
+                    nullable: false
+              - column:
+                  name: email
+                  type: VARCHAR(255)
+                  constraints:
+                    nullable: false
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/02-create-users-table.yaml:20]: add unique contraint

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/com/example/spring/config/MapperConfig.java [hunk 10]
```diff
@@ -6,8 +6,8 @@
 @org.mapstruct.MapperConfig(
         componentModel = "spring",
         injectionStrategy = InjectionStrategy.CONSTRUCTOR,
-        nullValueCheckStrategy = NullValueCheckStrategy.ALWAYS,
-        implementationName = "BookMapperImpl"
+        nullValueCheckStrategy = NullValueCheckStrategy.ALWAYS
+//        implementationName = "BookMapperImpl"
```
Comment (okuzan) [inline src/main/java/com/example/spring/config/MapperConfig.java]: remove comment

File: src/main/java/com/example/spring/exception/ErrorResponse.java [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+@Setter
+public class ErrorResponse {
+    private String error;
+    private String message;
+
+    public ErrorResponse(String error, String message) {
+        this.error = error;
+        this.message = message;
+    }
+}
```
Comment (okuzan) [inline src/main/java/com/example/spring/exception/ErrorResponse.java]: make this a record, or use @Data

File: src/main/java/com/example/spring/models/User.java [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+import lombok.Setter;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.SQLRestriction;
+
+@Entity
+@Table(name = "users")
+@Getter
+@Setter
+@SQLDelete(sql = "UPDATE users SET is_deleted = true WHERE id=?")
+@SQLRestriction(value = "is_deleted=false")
```
Comment (okuzan) [inline src/main/java/com/example/spring/models/User.java]:
```suggestion
@SQLDelete(sql = "UPDATE users SET is_deleted = TRUE WHERE id = ?")
@SQLRestriction(value = "is_deleted = FALSE")
```

**Review by okuzan — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/spring/exception/ErrorResponse.java [hunk 13]
```diff
@@ -0,0 +1,15 @@
+package com.example.spring.exception;
+
+import lombok.Data;
+
+@Data
+public class ErrorResponse {
+    private String error;
+    private String message;
+
+    public ErrorResponse(String error, String message) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/spring/exception/ErrorResponse.java]: remove
`@Data` provides constructor

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
one comment

**Review by Elena-Bruyako — APPROVED** [review]

**Pull request merged**